### PR TITLE
pkg/tasks: setup GRPC TLS for thanos querier

### PR DESCRIPTION
**Pull request by s-urbaniak**

### CODE REVIEW

File: pkg/manifests/tls.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+func getSelfSignedCA(s *v1.Secret) (*crypto.CA, error) {
+	crt, crtPresent := s.Data["ca.crt"]
+	key, keyPresent := s.Data["ca.key"]
+
+	if crtPresent && keyPresent {
+		ca, err := crypto.GetCAFromBytes(crt, key)
+		if err != nil {
+			klog.Warningf("creating a new CA due to error reading CA: %v", err)
+		}
+		if !needsNewCert(ca.Config.Certs[0].NotBefore, ca.Config.Certs[0].NotAfter, time.Now) {
```
Comment (s-urbaniak) [inline pkg/manifests/tls.go]: this needs a `if ca != nil`

Comment (s-urbaniak) [reply pkg/manifests/tls.go]: (or simply an `else if`)

File: pkg/manifests/manifests.go [hunk 2]
```diff
@@ -2586,6 +2650,42 @@ func (f *Factory) HashTrustedCA(caBundleCM *v1.ConfigMap, prefix string) *v1.Con
 	}
 }
 
+// HashSecret synthesizes a configmap just by copying "ca-bundle.crt" from the given configmap
```
Comment (s-urbaniak) [inline pkg/manifests/manifests.go]: fix this godoc

File: pkg/manifests/manifests.go [hunk 3]
```diff
@@ -2586,6 +2650,42 @@ func (f *Factory) HashTrustedCA(caBundleCM *v1.ConfigMap, prefix string) *v1.Con
...
+// and naming it by hashing the contents of "ca-bundle.crt".
+// It adds "monitoring.openshift.io/name" and "monitoring.openshift.io/hash" labels.
+// Any other labels from the given configmap are discarded.
+//
+// It returns nil, if the given configmap does not contain the "ca-bundle.crt" the data key
+// or data is empty string.
+func (f *Factory) HashSecret(secret *v1.Secret, data ...string) (*v1.Secret, error) {
```
Comment (s-urbaniak) [inline pkg/manifests/manifests.go]: add some unit tests

File: pkg/tasks/prometheus_user_workload.go [hunk 4]
```diff
@@ -139,8 +139,45 @@ func (t *PrometheusUserWorkloadTask) create() error {
 		return errors.Wrap(err, "reconciling UserWorkload Prometheus Service failed")
 	}
 
+	grpcTLS, err := t.factory.GRPCSecret(nil)
+	if err != nil {
+		return errors.Wrap(err, "initializing Prometheus GRPC secret failed")
```
Comment (s-urbaniak) [inline pkg/tasks/prometheus_user_workload.go]: fix error message: `s/Prometheus/UserWorkload Prometheus`

File: pkg/tasks/thanos_querier.go [hunk 5]
```diff
@@ -61,12 +61,12 @@ func (t *ThanosQuerierTask) Run() error {
 		return errors.Wrap(err, "waiting for Thanos Querier Route to become ready failed")
 	}
 
-	s, err := t.factory.ThanosQuerierOauthCookieSecret()
+	grpcTLS, err := t.factory.ThanosQuerierOauthCookieSecret()
```
Comment (s-urbaniak) [inline pkg/tasks/thanos_querier.go]: whops, this renamed a bit too much, rename it back to `s`

File: pkg/manifests/manifests.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/manifests/manifests.go]: done

File: pkg/manifests/manifests.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/manifests/manifests.go]: done

File: pkg/manifests/tls.go (hunk 1 above)
Comment (s-urbaniak) [reply pkg/manifests/tls.go]: done

File: pkg/tasks/prometheus_user_workload.go (hunk 4 above)
Comment (s-urbaniak) [reply pkg/tasks/prometheus_user_workload.go]: done

File: pkg/tasks/thanos_querier.go (hunk 5 above)
Comment (s-urbaniak) [reply pkg/tasks/thanos_querier.go]: done

**Comment by s-urbaniak** [comment]
@LiliC @paulfantom @brancz this is ready for a next pass

**Comment by s-urbaniak** [comment]
Let's get some more pair of eyes.

File: pkg/manifests/tls.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		s.Data = make(map[string][]byte)
+		s.Annotations = make(map[string]string)
+	}
+
+	ca, err := getSelfSignedCA(s)
+	if err != nil {
+		return nil, errors.Wrap(err, "error rotating CA")
+	}
+
+	// no need to rotate
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/manifests/tls.go]: good catch, let's add some godoc to `getSelfSignedCA`, it returns a `nil` CA if no rotation is necessary.

File: pkg/client/client.go [hunk 7]
```diff
@@ -592,6 +611,40 @@ func (c *Client) WaitForStatefulsetRollout(sts *appsv1.StatefulSet) error {
 	return nil
 }
 
+func (c *Client) WaitForSecret(s *v1.Secret) (*v1.Secret, error) {
+	var result *v1.Secret
+	var lastErr error
+	if err := wait.Poll(1*time.Second, 1*time.Minute, func() (bool, error) {
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/client/client.go]: hah, that's a left-over from initial testing :+1: I need to increase it to `5*time.Minute` as we do everywhere else.

File: pkg/manifests/tls.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+func getSelfSignedCA(s *v1.Secret) (*crypto.CA, error) {
+	crt, crtPresent := s.Data["ca.crt"]
+	key, keyPresent := s.Data["ca.key"]
+
+	if crtPresent && keyPresent {
+		ca, err := crypto.GetCAFromBytes(crt, key)
+		if err != nil {
+			klog.Warningf("creating a new CA due to error reading CA: %v", err)
+		} else if !needsNewCert(ca.Config.Certs[0].NotBefore, ca.Config.Certs[0].NotAfter, time.Now) {
+			return nil, nil
```
Comment (brancz) [inline pkg/manifests/tls.go]: I feel like this check rather belongs in the calling function, I had to jump there and back to find the 

```
// no need to rotate
```

comment in the other function. At the very least we should comment this function with what the `nil` return means.

File: pkg/manifests/tls.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+			return nil, errors.Wrap(err, "error getting PEM bytes for thanos querier client certificate")
+		}
+		s.Data["thanos-querier-client.crt"] = crt
+		s.Data["thanos-querier-client.key"] = key
+	}
+
+	{
+		cfg, err := ca.MakeServerCert(
+			sets.NewString("prometheus-grpc"),
+			crypto.DefaultCertificateLifetimeInDays,
```
Comment (brancz) [inline pkg/manifests/tls.go]: This one doesn't need the `*24*time.Hour` like the other method?

File: pkg/manifests/tls.go (hunk 8 above)
Comment (s-urbaniak) [reply pkg/manifests/tls.go]: agreed, good catch. I will refactor for better readability.

File: pkg/manifests/tls.go (hunk 9 above)
Comment (s-urbaniak) [reply pkg/manifests/tls.go]: No, in that case the library expects an `int`: https://github.com/openshift/cluster-monitoring-operator/blob/c6195af44d80d4358f15f8e015c6cc608326f224/vendor/github.com/openshift/library-go/pkg/crypto/crypto.go#L665:15

Comment (brancz) [reply pkg/manifests/tls.go]: Ack

**Comment by s-urbaniak** [comment]
@LiliC @brancz @paulfantom ptal

File: pkg/manifests/manifests.go [hunk 10]
```diff
@@ -2610,6 +2674,45 @@ func (f *Factory) HashTrustedCA(caBundleCM *v1.ConfigMap, prefix string) *v1.Con
 	}
 }
 
+// HashSecret synthesizes a secret by setting the given data
```
Comment (bwplotka) [inline pkg/manifests/manifests.go]: Hmm this sentence sounds odd to me ):

File: jsonnet/thanos-querier.jsonnet [hunk 11]
```diff
@@ -177,29 +183,33 @@ local authorizationRole =
...
                       exec: {
-                        command: ["sh", "-c", "curl http://localhost:9090/-/healthy"],
+                        command: ['sh', '-c', 'curl http://localhost:9090/-/healthy'],
                       },
                     },
                     readinessProbe: {
                       httpGet:: {},
                       exec: {
-                        command: ["sh", "-c", "curl http://localhost:9090/-/healthy"],
+                        command: ['sh', '-c', 'curl http://localhost:9090/-/healthy'],
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply jsonnet/thanos-querier.jsonnet]: hmm .. does thanos querier have a dedicated `/-/ready` endpoint? :thinking: i think i took this over from observatorium, cc @metalmatze

File: pkg/client/client.go [hunk 12]
```diff
@@ -592,6 +611,40 @@ func (c *Client) WaitForStatefulsetRollout(sts *appsv1.StatefulSet) error {
 	return nil
 }
 
+func (c *Client) WaitForSecret(s *v1.Secret) (*v1.Secret, error) {
+	var result *v1.Secret
+	var lastErr error
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/client/client.go]: agreed, i will add it (we use that method/idiom quite often though in CMO in other methods too).

File: pkg/manifests/manifests.go (hunk 10 above)
Comment (s-urbaniak) [reply pkg/manifests/manifests.go]: indeed :-) I will change this as a follow up.

**Pull request merged**